core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java

```diff
@@ -171,4 +175,19 @@ public void close() {
     void doClose() {
         clientFactoriesToClose.forEach(ClientFactory::close);
     }
+
+    private URI normalizeUri(URI uri, Scheme scheme) {
+        if (isUndefinedUri(uri)) {
+            // See https://github.com/line/armeria/pull/2241 discussion for more details.
+            return uri;
+        }
+        if (scheme.serializationFormat() != SerializationFormat.NONE) {
+            return uri;
+        }
+
+        final String newScheme = uri.getScheme()
```

Comment:
I believe it's more clear to use `scheme.sessionProtocol()` instead of string replacement - we are always replacing something like `none+http` or `https+none` with the protocol so may as well write the code like that.

Might also consider using the non-string constructor - I think it's more clear since there's no thinking about "is `substring` length actually correct, will parse throw errors", etc, it's clearly doing exactly what we want to do, replacing the scheme.

```java
new URI(scheme.sessionProtocol().uriText(),
             u.getUserInfo(), u.getHost(), u.getPort(),
             u.getPath(), u.getQuery(),
             u.getFragment())
```

Reply:
I'd prefer using a single-string constructor because multi-string constructors concatenates the components and parses it anyway.

Reply:
@trustin @anuraaga : Sorry, I am a bit confused. any change is needed for this? Without going into URI construction complexities, this one tackles the scheme only. 
WDYT?  

```
final String newScheme = uri.getScheme()
                                    .replace(SerializationFormat.NONE.uriText(), "")
                                    .replace("+", "");
        return URI.create(newScheme + uri.toString().substring(uri.getScheme().length()));
```

Reply:
I personally still like the multi-string constructor since it's indeed implemented similar to `URI.create(String)` but makes very clear which part we're replacing.

Otherwise, if going with the single string, I'd still recommend being explicit `scheme.sessionProtocol() + "://" + uri.substring(`. There doesn't seem to be a good reason to rely on the coincidence that session protocols never have the word none inside them, i.e., the current code wouldn't work for a scheme like `none+httpnone://google.com`. It's a contrived example, but I think shows that the current logic is more indirect than it needs to be.

Reply:
Agree. I have updated it to use
```
        return URI.create(scheme.sessionProtocol() + "://" +
                          uri.toString().substring(uri.getScheme().length()));
```